Content.Server/Explosion/EntitySystems/ExplosionSystem.Processing.cs

```diff
@@ -0,0 +1,647 @@
+
+        while (_robustRandom.Prob(type.TileBreakChance(intensity)))
+        {
+            intensity -= type.TileBreakRerollReduction;
+
+            if (tileDef is not ContentTileDefinition contentTileDef)
+                break;
+
+            // does this have a base-turf that we can break it down to?
+            if (contentTileDef.BaseTurfs.Count == 0)
+                break;
+
+            tileDef = _tileDefinitionManager[contentTileDef.BaseTurfs[^1]];
+        }
+
+        if (tileDef.TileId == tileRef.Tile.TypeId)
+            return;
+
+        damagedTiles.Add((tileRef.GridIndices, new Tile(tileDef.TileId)));
+    }
+}
+
+/// <summary>
+///     This is a data class that stores information about the area affected by an explosion, for processing by <see
+///     cref="ExplosionSystem"/>.
+/// </summary>
+/// <remarks>
+///     This is basically the output of <see cref="ExplosionSystem.GetExplosionTiles()"/>, but wrapped in an enumerator
+///     to iterate over the tiles, along with the ability to keep track of what entities have already been damaged by
+///     this explosion.
+/// </remarks>
+sealed class Explosion
+{
+    struct ExplosionData
+    {
+        public EntityLookupComponent Lookup;
+        public Dictionary<int, List<Vector2i>> TileLists;
+        public IMapGrid? MapGrid;
+    }
+
+    /// <summary>
+    ///     Used to avoid applying explosion effects repeatedly to the same entity. Particularly important if the
+    ///     explosion throws this entity, as then it will be moving while the explosion is happening.
+    /// </summary>
+    public readonly HashSet<EntityUid> ProcessedEntities = new();
+
+    /// <summary>
+    ///     This integer tracks how much of this explosion has been processed.
+    /// </summary>
+    public int CurrentIteration { get; private set; } = 0;
+
+    public readonly ExplosionPrototype ExplosionType;
+    public readonly MapCoordinates Epicenter;
+    private readonly Matrix3 _spaceMatrix;
+    private readonly Matrix3 _invSpaceMatrix;
+
+    private readonly List<ExplosionData> _explosionData = new();
+    private readonly List<float> _tileSetIntensity;
+
+    public bool FinishedProcessing;
+
+    // shitty enumerator implementation
+    private DamageSpecifier _currentDamage = default!;
+    private EntityLookupComponent _currentLookup = default!;
+    private IMapGrid? _currentGrid;
+    private float _currentIntensity;
+    private float _currentThrowForce;
+    private List<Vector2i>.Enumerator _currentEnumerator;
+    private int _currentDataIndex;
+    private Dictionary<IMapGrid, List<(Vector2i, Tile)>> _tileUpdateDict = new();
+
+    private EntityQuery<TransformComponent> _xformQuery;
+    private EntityQuery<PhysicsComponent> _physicsQuery;
+    private EntityQuery<DamageableComponent> _damageQuery;
+
+    public int Area;
+
+    private readonly ExplosionSystem _system;
+
+    public Explosion(ExplosionSystem system,
```

Comment:
the next couple functions in general could do with some more comments

Reply:
will add